File: python/enolib/elements/value_element_base.py
from ..errors.validation import Validation
from .element_base import ElementBase


class ValueElementBase(ElementBase):
    def _print_value(self):
        value = self._context.value(self._instruction)

        if not value:
            return 'None'

        if len(value) > 14:
            value = f"{value[0:11]}..."

        return value.replace("\n", '\n')

    def _value(self, loader=None, *, required):
        self._touched = True

        value = self._context.value(self._instruction)

        if value is None:
            if required:
                raise Validation.missing_value(self._context, self._instruction)
            else:
                return None

        # TODO: Check if ValueError is 1) a valid pythonic approach 2) the best choice

        if not loader:
            return value

        try:
            return loader(value)
        except ValueError as message:
            raise Validation.value_error(self._context, message, self._instruction)

    def optional_string_value(self):
        return self._value(required=False)

    def optional_value(self, loader):
        return self._value(loader, required=False)

    def required_string_value(self):
        return self._value(required=True)

    def required_value(self, loader):
        return self._value(loader, required=True)

    def value_error(self, message):
        if callable(message):
            message = message(self._context.value(self._instruction))

        return Validation.value_error(self._context, message, self._instruction)

File: ruby/lib/enolib/elements/value_element_base.rb
# frozen_string_literal: true

module Enolib
  class ValueElementBase < ElementBase
    def optional_string_value
      _value(required: false)
    end

    def optional_value(loader = nil, &block)
      loader = Proc.new(&block) if block_given?

      unless loader
        raise ArgumentError, 'A loader function must be provided'
      end

      _value(loader, required: false)
    end

    def required_string_value
      _value(required: true)
    end

    def required_value(loader = nil, &block)
      loader = Proc.new(&block) if block_given?

      unless loader
        raise ArgumentError, 'A loader function must be provided'
      end

      _value(loader, required: true)
    end

    def value_error(message = nil)
      if block_given?
        message = yield(@context.value(@instruction))
      elsif message.is_a?(Proc)
        message = message.call(@context.value(@instruction))
      end

      unless message
        raise ArgumentError, 'A message or message function must be provided'
      end

      Errors::Validation.value_error(@context, message, @instruction)
    end

    private

    def print_value
      value = @context.value(@instruction)

      return 'nil' unless value

      value = "#{value[0..10]}..." if value.length > 14

      value.gsub("\n", '\n')
    end

    def _value(loader = nil, required:)
      @touched = true

      value = @context.value(@instruction)

      if value
        return value unless loader

        begin
          loader.call(value)
        rescue => message
          raise Errors::Validation.value_error(@context, message, @instruction)
        end
      else
        return nil unless required

        raise Errors::Validation.missing_value(@context, @instruction)
      end
    end
  end
end
